### PR TITLE
[Wasm64] Fix EM_ASM for addresses >4GB addresses

### PR DESCRIPTION
**Pull request by sbc100**
This change use makeGetValue helper macros to avoid direct heap access.
    
This does come with a 6 byte code size cost (I measured the size of the
JS code generated for wasm64.test_em_asm with -O2 going from 9275 to
9281).

### CODE REVIEW

File: test/runner.py [hunk 1]
```diff
@@ -284,7 +284,7 @@ def load_test_suites(args, modules):
       for test in tests:
         suite.addTest(test)
       suites.append((m.__name__, suite))
-  if total_tests == 1:
+  if total_tests == 1 or parallel_testsuite.num_cores() == 1:
```
Comment (kripken) [inline test/runner.py:287]: Is this related?

**Review by kripken — COMMENTED** [review]
I understand this uses makeGetValue to make sure to emit the right code, but what was breaking in the old code? We do support direct `HEAPF64[..]` etc. access in wasm64, don't we? And `>>=` should work on an int53?

File: test/runner.py (hunk 1 above)
Comment (sbc100) [reply test/runner.py:287]: Yes, when running 4gb tests I run with `--cores=1` in which case I want use save the test results in `out/test` .. just like if we run a single test

**Comment by sbc100** [comment]
> I understand this uses makeGetValue to make sure to emit the right code, but what was breaking in the old code? We do support direct `HEAPF64[..]` etc. access in wasm64, don't we? And `>>=` should work on an int53?

I believe the problem with using `buf >>= 2` to get to turn the byte address into an int32 address.  This doesn't work if buf is largers than 2**32, right?

**Comment by sbc100** [comment]
> > I understand this uses makeGetValue to make sure to emit the right code, but what was breaking in the old code? We do support direct `HEAPF64[..]` etc. access in wasm64, don't we? And `>>=` should work on an int53?
> 
> I believe the problem with using `buf >>= 2` to get to turn the byte address into an int32 address. This doesn't work if buf is largers than 2**32, right?

Indeed any usage of bitshifting doesn't work for these larger addresses, right?

**Comment by kripken** [comment]
Ah, yes, that does make sense. We'd need `>>>=` for 2-4GB but nothing can work for more than that.

Another option might be `/= 2`. The JIT will make it fast anyhow, and it would work up to 53 bits. What do you think?

**Comment by sbc100** [comment]
> Ah, yes, that does make sense. We'd need `>>>=` for 2-4GB but nothing can work for more than that.
> 
> Another option might be `/= 2`. The JIT will make it fast anyhow, and it would work up to 53 bits. What do you think?

Whats wrong with the approach taken here which uses the getVale/setValue helpers to avoid this completely?  In most cases I think its more consistent to always use these helpers rather than direct heap access, no?

**Comment by sbc100** [comment]
> Ah, yes, that does make sense. We'd need `>>>=` for 2-4GB but nothing can work for more than that.
> 
> Another option might be `/= 2`. The JIT will make it fast anyhow, and it would work up to 53 bits. What do you think?

Whats wrong with the approach taken here which uses the getVale/setValue helpers to avoid this completely?  In most cases I think its more consistent to always use these helpers rather than direct heap access, no?

**Comment by kripken** [comment]
This approach is good. It's just the 11 byte code cost. Not a big deal but if this pattern happens in many places perhaps it adds up, that's what worries me. And I don't see obvious downsides to `/=`?

**Comment by sbc100** [comment]
OK I switched to using division, which actually saves a few bytes.   How confident are you the optimizing compiler can make these equivalent?

**Comment by sbc100** [comment]
@juj has another solution that involves yet more macros: https://github.com/emscripten-core/emscripten/pull/19289

**Comment by kripken** [comment]
I'm very confident VMs optimize this based on what I recall from years ago, but my information isn't recent, so we should probably verify to make sure.

If we verify, this PR lgtm, but good point about that other PR with the macros, perhaps that is worth considering as well - what do you think? I lean towards just doing `/ *` actually, since it seems simpler (and even saves size compared to `>> <<0.

**Comment by sbc100** [comment]
I reversed course again and decided to go with the helper macros here.. I think its cleaner and more consistent.  And worth the 6 bytes of code size.   It also unblocks a lot of followup changes to get >4gb tests passing.

**Comment by sbc100** [comment]
We can look into the possibility of using division rather than shifting if/when we have some performance numbers, but using the macros means that we can update in single location, making these measurements easier.

**Review by kripken — APPROVED** [review]
lgtm. if we find this extra code size appears in more places we can maybe prioritize focusing on it more then. if it's just a few places then the simplicity and other benefits seem worth it.
